types/react-beautiful-dnd/index.d.ts

```diff
@@ -37,70 +180,121 @@ export interface DropResult {
+    position: 'fixed';
 
-export interface DroppableProvided {
-    innerRef(element: HTMLElement | null): any;
-    placeholder?: React.ReactElement<any>;
+    // When we do `position: fixed` the element looses its normal dimensions,
+    // especially if using flexbox. We set the width and height manually to
+    // ensure the element has the same dimensions as before it started dragging
+    width: number;
+    height: number;
+
+    // The width and height values take into account whether the original element
+    // used `box-sizing: content-box` or `box-sizing: border-box`
+    // Because we are setting the width and height directly we want to ensure that
+    // these are the actual values applied
+    boxSizing: 'border-box';
+
+    // We initially position the element in the same *visual spot* as when it started.
+    // This means that these values *exclude* the original margins so that element remains
+    // in the same visual position - even though now it has no margins
+    top: number;
+    left: number;
+
+    // We clear any top or left margins on the element to ensure it does not push
+    // the element positioned with the top/left position (which is margin aware).
+    // We also clear the margin right / bottom. This has no positioning impact,
+    // but it is cleanest to just remove all the margins rather than only the top and left.
+    margin: 0;
+
+    // We need to opt out of the shared global style that is being applied to
+    // all draggables. The movement of moving draggables is either not animated
+    // or handled by react-motion.
+    transition: 'none';
+
+    // Move the element in response to a user dragging
+    transform?: string;
+
+    // When dragging or dropping we control the z-index to ensure that
+    // the layering is correct
+    zIndex: ZIndex;
 }
 
-export interface DroppableStateSnapshot {
-    isDraggingOver: boolean;
+export interface NotDraggingStyle {
+    transform?: string;
+    // null: use the global animation style
+    // none: skip animation (used in certain displacement situations)
+    transition: null | 'none';
 }
 
-export interface DroppableProps {
-    droppableId: DroppableId;
-    type?: TypeId;
-    isDropDisabled?: boolean;
-    direction?: 'vertical' | 'horizontal';
-    children(provided: DroppableProvided, snapshot: DroppableStateSnapshot): React.ReactElement<any>;
+export type DraggableStyle = DraggingStyle | NotDraggingStyle;
+
+// Props that can be spread onto the element directly
+export interface DraggableProps {
+    // inline style
+    style?: DraggableStyle;
+    // used for shared global styles
+    'data-react-beautiful-dnd-draggable': string;
 }
 
-export class Droppable extends React.Component<DroppableProps> {}
+export interface DraggableProvided {
+    draggableProps: DraggableProps;
+    // will be null if the draggable is disabled
+    dragHandleProps?: DragHandleProps;
```

Comment:
This should probably be `dragHandleProps: DragHandleProps | undefined | null;`

Remember that [Flow's Maybe type](https://flow.org/en/docs/types/maybe/) `?Foo` translates to `Foo | undefined | null` in TypeScript.

There are a few other lines in this PR that will need this change applied too.

Reply:
The same class of bug [exists in the current definitions too](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/22938#issuecomment-357844824), so it would be great to fix this in 4.0

Reply:
OK, I am seeing a lot of usage of `null` in the source code so there is a chance that some of these values will indeed be null at some point. Will introduce `type Maybe<T> = T | null | undefined` and replace the relevant parts with `Maybe<T>`.

But I wonder, what to do with `Droppable_MapProps` and `Draggable_MapProps` - I had to make them optional because even though they are part of the Internal props, they are not to be specified by the user. I believe those will not be `null`. Does it make sense?

Reply:
In 3.0 there were definitely nulls (I had bugs because of it!), so the `Maybe` type sounds good. Regarding `Droppable_MapProps` and `Draggable_MapProps`, it's fine for properties to be both optional and `Maybe`, e.g. `foo?: Maybe<Foo>`

Reply:
Further to this, I'd be interested to know if Flow's Maybe actually translates more accurately to `foo?: Foo | undefined | null` or `foo: Foo | undefined | null`. I'm leaning towards the former.

Reply:
I just pushed a change that turns all the instances of `:? T` from the original flow code into `: Maybe<T>` in ts code.
But for all the other cases that were typed explicitly in flow and that I turned into optional to respect the documented API I left as `?: T`. Not it is easier to make a distinction between those cases.

At the same time, can you clarify how does it come that in flow these cases were working fine? For example, props `isDragDisabled` and `disableInteractiveElementBlocking` from `Draggable_OwnProps`.